src/rust/zh.mangabz/src/parser.rs

```diff
@@ -0,0 +1,296 @@
+use aidoku::{
+	error::Result,
+	helpers::substring::Substring,
+	prelude::format,
+	std::{html::Node, net::Request, String, Vec},
+	Chapter, Filter, FilterType, Manga, MangaPageResult, MangaStatus, Page,
+};
+
+extern crate alloc;
+use alloc::string::ToString;
+
+pub const BASE_URL: &str = "https://mangabz.com/";
+pub const USER_AGENT: &str = "Mozilla/5.0 (Macintosh; Intel Mac OS X 13_3_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/113.0.0.0 Safari/537.36";
+const GENRE: [u8; 10] = [0, 31, 26, 1, 2, 25, 11, 17, 15, 34];
+const SORT: [u8; 2] = [10, 2];
+
+pub fn get_filtered_url(filters: Vec<Filter>, page: i32) -> String {
+	let mut url = String::from(BASE_URL);
+
+	let mut is_searching = false;
+	let mut search_str = String::new();
+
+	let mut genre = 0;
+	let mut status = 0;
+	let mut sort = 10;
+
+	for filter in filters {
+		match filter.kind {
+			FilterType::Title => {
+				if let Ok(filter_value) = filter.value.as_string() {
+					is_searching = true;
+					search_str = filter_value.read();
+				}
+			}
+			FilterType::Select => {
+				let index = filter.value.as_int().unwrap_or(0) as usize;
+				match filter.name.as_str() {
+					"題材" => genre = GENRE[index],
+					"狀態" => status = index as u8,
+					_ => continue,
+				}
+			}
+			FilterType::Sort => {
+				let value = match filter.value.as_object() {
+					Ok(value) => value,
+					Err(_) => continue,
+				};
+				let index = value.get("index").as_int().unwrap_or(0) as usize;
+				sort = SORT[index];
+			}
+			_ => continue,
+		}
+	}
+
+	if is_searching {
+		url.push_str(format!("search?title={}&page={}", search_str, page).as_str());
+	} else {
+		url.push_str(format!("manga-list-{}-{}-{}-p{}/", genre, status, sort, page).as_str());
+	}
+
+	url
+}
+
+pub fn request_get(url: String) -> Request {
+	Request::get(url)
+		.header("Referer", BASE_URL)
+		.header("User-Agent", USER_AGENT)
+}
+
+pub fn get_manga_list(html: Node) -> Result<MangaPageResult> {
+	let mut mangas: Vec<Manga> = Vec::new();
+
+	for item in html.select("div.mh-item").array() {
+		let manga_item = item.as_node()?;
+		let title_node = manga_item.select("h2.title > a");
+
+		let id = title_node
+			.attr("href")
+			.read()
+			.replace('/', "")
+			.replace("bz", "");
+		let cover = manga_item.select("img.mh-cover").attr("src").read();
+		let title = title_node.attr("title").read();
+		let url = format!("{}{}bz/", BASE_URL, id);
+
+		let status_str = manga_item.select("span").text().read();
+		let status = match status_str.as_str() {
+			"最新" => MangaStatus::Ongoing,
+			"完結" => MangaStatus::Completed,
+			_ => MangaStatus::Unknown,
+		};
+
+		mangas.push(Manga {
+			id,
+			cover,
+			title,
+			url,
+			status,
+			..Default::default()
+		});
+	}
+
+	let has_more = !html
+		.select("div.page-pagination a:contains(>)")
+		.array()
+		.is_empty();
+
+	html.close();
```

Comment:
Just wanna ask, should I close requests and nodes manually, or the system would do it for me?

Reply:
the aidoku rust wrapper makes use of rust's ownership system, so as soon as the request object exits scope it's closed automatically, so it's best to not do it manually

Reply:
ok, thanks!